cmd/kubectl.go

```diff
@@ -37,34 +48,66 @@ func runKubectl(ctx *cli.Context) error {
 		return err
 	}
 
-	cluster, err := getClusterByID(c, c.UserConfig.FocusedCluster())
+	config, err := loadConfig(ctx)
 	if err != nil {
 		return err
 	}
 
-	config, err := c.ManagementClient.Cluster.ActionGenerateKubeconfig(cluster)
-	if err != nil {
-		return err
+	currentRancherServer := config.FocusedServer()
+	if currentRancherServer == nil {
+		return fmt.Errorf("no focused server")
 	}
 
-	tmpfile, err := ioutil.TempFile("", "rancher-")
+	currentToken := currentRancherServer.AccessKey
+	t, err := c.ManagementClient.Token.ByID(currentToken)
 	if err != nil {
 		return err
 	}
-	defer os.Remove(tmpfile.Name())
 
-	_, err = tmpfile.Write([]byte(config.Config))
+	currentUser := t.UserID
+	currentCluster := currentRancherServer.FocusedCluster()
+	filename := fmt.Sprintf("%s-%s-kubeconfig", currentUser, currentCluster)
+	home, err := os.UserHomeDir()
 	if err != nil {
 		return err
 	}
+	kubeconfigFilePath := fmt.Sprintf("%s/%s/%s", home, kubeconfigpath, filename)
 
-	err = tmpfile.Close()
+	kubeconfig, err := loadKubeconfig(kubeconfigFilePath)
 	if err != nil {
 		return err
 	}
 
+	var isTokenValid bool
+	if len(kubeconfig) != 0 {
+		tokenID, err := extractKubeconfigTokenID(kubeconfig)
+		if err != nil {
+			return err
+		}
+		isTokenValid, err = validateToken(tokenID, c.ManagementClient.Token)
+		if err != nil {
+			return err
+		}
+	}
+
+	if len(kubeconfig) == 0 || !isTokenValid {
+		cluster, err := getClusterByID(c, c.UserConfig.FocusedCluster())
+		if err != nil {
+			return err
+		}
+
+		config, err := c.ManagementClient.Cluster.ActionGenerateKubeconfig(cluster)
+		if err != nil {
+			return err
+		}
+		kubeconfig = []byte(config.Config)
+		if err := writeKubeconfig(kubeconfigFilePath, kubeconfig); err != nil {
+			return err
+		}
+	}
+
 	cmd := exec.Command(path, ctx.Args()...)
-	cmd.Env = append(os.Environ(), "KUBECONFIG="+tmpfile.Name())
+	cmd.Env = append(os.Environ(), "KUBECONFIG="+kubeconfigFilePath)
```

Comment:
Should we be using [`os.Setenv`](https://golang.org/pkg/os/#Setenv) here?

Reply:
I think this way may be better since the var is only needed for the single command and this contains it.

Reply:
Oh sorry you're right. I didn't realize this was just being set for this command.